File: api/utils/ordering.go
package utils

import "strings"

const alphabet = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"

const midChar = "V"

func GenerateNextKey(prev string) string {
	if prev == "" {
		return "a"
	}

	lastChar := prev[len(prev)-1]
	idx := strings.IndexByte(alphabet, lastChar)

	if idx == -1 || idx == len(alphabet)-1 {
		return prev + string(alphabet[0])
	}

	return prev[:len(prev)-1] + string(alphabet[idx+1])
}

func GenerateKeyBetween(prev, next *string) string {
	if prev == nil {
		if next == nil {
			return "a"
		}
		return midpoint("", *next)
	}

	if next == nil {
		return GenerateNextKey(*prev)
	}

	if *prev >= *next {
		return GenerateNextKey(*prev)
	}

	return midpoint(*prev, *next)
}

func midpoint(prev, next string) string {
	pLen := len(prev)
	nLen := len(next)
	i := 0

	for i < pLen && i < nLen && prev[i] == next[i] {
		i++
	}

	valP := -1
	if i < pLen {
		valP = strings.IndexByte(alphabet, prev[i])
	} else {
		return prev + midChar
	}

	valN := strings.IndexByte(alphabet, next[i])

	if valP == -1 || valN == -1 {
		return prev + midChar
	}

	if valN-valP > 1 {
		mid := (valP + valN) / 2
		return prev[:i] + string(alphabet[mid])
	}

	return prev + midChar
}

File: app/lib/core/utils/ordering.dart
String generateNextKey(String prev) {
  if (prev.isEmpty) return 'a';

  int lastCode = prev.codeUnitAt(prev.length - 1);

  if (lastCode == 122) {
    return '${prev}a';
  }

  return prev.substring(0, prev.length - 1) + String.fromCharCode(lastCode + 1);
}

String generateKeyBetween(String? prev, String? next) {
  if (prev == null || prev.isEmpty) {
    if (next == null || next.isEmpty) return 'a';
    return _midpoint("", next);
  }
  if (next == null || next.isEmpty) {
    return _nextKey(prev);
  }

  return _midpoint(prev, next);
}

String _midpoint(String prev, String next) {
  int pLen = prev.length;
  int nLen = next.length;
  int i = 0;

  while (i < pLen && i < nLen && prev.codeUnitAt(i) == next.codeUnitAt(i)) {
    i++;
  }

  int charP = (i < pLen) ? prev.codeUnitAt(i) : 48;

  int charN = (i < nLen) ? next.codeUnitAt(i) : 123;

  if (charN - charP > 1) {
    int midCode = (charP + charN) ~/ 2;
    return prev.substring(0, i) + String.fromCharCode(midCode);
  }

  return '${prev}n';
}

String _nextKey(String prev) {
  int lastCode = prev.codeUnitAt(prev.length - 1);
  if (lastCode < 122) {
    return prev.substring(0, prev.length - 1) +
        String.fromCharCode(lastCode + 1);
  }
  return '${prev}a';
}
